src/Jackalope/Version/VersionHandler.php

```diff
@@ -119,27 +128,52 @@ public function checkinItem($path)
 
         // TODO set subgraph to read only
 
-        $versionHistoryNode = $node->getPropertyValue('jcr:versionHistory');
+        /** @var NodeInterface $versionHistoryNode */
+        $versionHistoryUuid = $node->getProperty('jcr:versionHistory')->getString();
+        $versionHistoryNode = $this->objectManager->getNodeByIdentifier($versionHistoryUuid, 'Version\VersionHistory');
 
         // FIXME add some kind of sharding
         // should avoid to have too many nodes on the same level
         $versionNode = $versionHistoryNode->addNode(UUIDHelper::generateUUID(), 'nt:version');
         $versionNode->setProperty('jcr:uuid', UUIDHelper::generateUUID());
         $versionNode->setProperty('jcr:created', new \DateTime());
-
-        // TODO create frozen node
-
-        $versionNode->setProperty('jcr:predecessors', $node->getProperty('jcr:predecessors')->getString());
-        $node->setProperty('jcr:predecessors', array(), PropertyType::REFERENCE, false);
-        foreach ($versionNode->getProperty('jcr:predecessors') as $predecessorUuid) {
-            $predecessor = $this->objectManager->getNodeByIdentifier($predecessorUuid);
-            $predecessor->setProperty('jcr:successors', array_merge($predecessor->getPropertyValueWithDefault('jcr:successors', array()), array($versionNode)));
+        $versionNode->setProperty('jcr:successors', array());
+
+        $this->createFrozenNode($versionNode, $node);
+
+        $baseVersionNode->setProperty('jcr:successors', array($versionNode), PropertyType::REFERENCE, false);
+        $baseVersionNode->setModified();
+        $versionNode->setProperty(
+            'jcr:predecessors',
+            $node->getProperty('jcr:predecessors')->getString(),
+            PropertyType::REFERENCE
+        );
+
+        $node->setProperty(
+            'jcr:predecessors',
+            array($versionNode),
```

Comment:
What do you mean by all? I think you can only have more than one predecessor if you have branching and merging, which is currently not implemented.